src/Storages/StorageMergeTree.cpp

```diff
@@ -1428,72 +1409,288 @@ MergeTreeDataPartPtr StorageMergeTree::outdatePart(MergeTreeTransaction * txn, c
+            throw Exception(ErrorCodes::PART_IS_TEMPORARILY_LOCKED,
+                            "Race with concurrent query that modifies parts. {} parts have changed the status, first is {}. Try again later.",
+                            changed_parts.size(), changed_parts.front()->getNameWithState());
+    }
+
+    DataPartsVector covered_parts;
+
+    for (auto part: new_parts)
+    {
+        auto builder = part->data_part_storage->getBuilder();
+        bool no_covering_parts = renameTempPartAndReplaceUnlocked(part, transaction, builder, part_lock, &covered_parts);
+        bool has_covering_parts = !no_covering_parts;
+        if (has_covering_parts)
+            throw Exception("Part " + part->name + " has covering part. This is a bug.", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    if (covered_parts.size() != old_parts.size())
+        throw Exception(ErrorCodes::LOGICAL_ERROR,
+                        "Going to delete {} (first is  {}) parts instead of {} (first is {}) by creating new {} (first is {}) parts. This is a bug.",
+                        covered_parts.size(), covered_parts.front()->getNameWithState(),
+                        old_parts.size(), old_parts.front()->getNameWithState(),
+                        new_parts.size(), new_parts.front()->getNameWithState()
+                        );
+
+    LOG_INFO(log, "Remove {} parts by covering them with empty {} parts.", covered_parts.size(), new_parts.size());
+
+    // Do commit at the same locked scope where new parts was checked
+    transaction.commit(&part_lock);
+
+    if (deduplication_log)
+        for (const auto & part: covered_parts)
             deduplication_log->dropPart(part->info);
+}
+
+void StorageMergeTree::truncate(const ASTPtr &, const StorageMetadataPtr &, ContextPtr query_context, TableExclusiveLockHolder &)
+{
+    /// Asks to complete merges and does not allow them to start.
+    /// This protects against "revival" of data for a removed partition after completion of merge.
+    auto merge_blocker = stopMergesAndWait();
+
+    Stopwatch watch;
+
+    auto txn = query_context->getCurrentTransaction();
+    MergeTreeData::Transaction transaction(*this, txn.get());
+    {
+        DataPartsVector parts = getVisibleDataPartsVector(query_context);
+
+        MutableDataPartsVector new_parts = makeCoveringEmptyTmpParts(parts, txn);
+
+        coverPartsWithEmptyParts(parts, new_parts, transaction);
+
+        PartLog::addNewParts(query_context, new_parts, watch.elapsed());
+
+        LOG_INFO(log, "Truncated table with {} parts by replacing them with new empty {} parts.",
+                 parts.size(), new_parts.size());
     }
 
-    if (detach)
-        LOG_INFO(log, "Detached {} parts.", parts_to_remove.size());
-    else
-        LOG_INFO(log, "Removed {} parts.", parts_to_remove.size());
+    /// Old parts are needed to be destroyed before clearing them from filesystem.
+    clearOldMutations(true);
+    clearOldPartsFromFilesystem();
```

Comment:
We can clear old parts forcefully to avoid modifying `old_parts_lifetime` in tests (see `IMergeTreeDataPart::remove_time`)

Reply:
Will look